File: Java/com/leetcode/removeDuplicatesFromSortedListII/RemoveDuplicatesFromSortedListII.java
package com.leetcode.removeDuplicatesFromSortedListII;

import com.leetcode.dataStructures.ListNode;

public class RemoveDuplicatesFromSortedListII {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode dummy = new ListNode(-1);
        ListNode prev = dummy;
        ListNode current = head;
        
        dummy.next = head;
        
        while (current != null) {
            // In the duplicate block
            if (current.next != null && current.val == current.next.val) {
                while (current.next != null && current.val == current.next.val) {
                    current = current.next;
                }
                
                if (current.next != null) {
                    current = current.next;
                    prev.next = current;
                } else {
                    prev.next = null;
                }
            } else {
                prev = current;
                current = current.next;
            }
        }
        
        return dummy.next;

    }
}

File: Python/remove_duplicates_from_sorted_list_ii/remove_duplicates_from_sorted_list_ii.py
class Solution(object):
	def deleteDuplicates(self, head):
		if head is None:
			return head
		# 1 -> 1
		dummy = ListNode(0)
		dummy.next = head
		p = dummy
		d = False
		while p.next and p.next.next:
			if p.next.val == p.next.next.val:
				p.next.next = p.next.next.next
				d = True
				continue
			if d:
				p.next = p.next.next
				d = False
			else:
				p = p.next
		if d:
			p.next = p.next.next
		return dummy.next
